runtime/src/accounts_index.rs

```diff
@@ -382,4 +383,25 @@ mod tests {
         assert_eq!(num, 1);
         assert!(found_key);
     }
+
+    #[test]
+    fn test_purge() {
```

Comment:
It seems test is not enough:

```
ryoqun@ubuqun:~/work/solana/solana$ git diff
diff --git a/runtime/src/accounts_db.rs b/runtime/src/accounts_db.rs
index ac7a15c22..5b1f19c31 100644
--- a/runtime/src/accounts_db.rs
+++ b/runtime/src/accounts_db.rs
@@ -607,7 +607,7 @@ impl AccountsDB {
         for pubkey in purges.keys() {
             let purge = accounts_index.purge(&pubkey);
             if purge.1 {
-                dead_keys.push(*pubkey);
+                //dead_keys.push(*pubkey);
             }
             reclaims.extend(purge.0);
         }
ryoqun@ubuqun:~/work/solana/solana$ (export CHANNEL=on-$(whoami)-fork CI_COMMIT=$(git rev-parse HEAD) && RUST_LOG=info RUST_BACKTRACE=1 RUST_BACKTRACE=full nice -n 20 cargo test --lib -p solana-runtime test_purge)
warning: variable does not need to be mutable
   --> runtime/src/accounts_db.rs:606:13
    |
606 |         let mut dead_keys = Vec::new();
    |             ----^^^^^^^^^
    |             |
    |             help: remove this `mut`
    |
    = note: `#[warn(unused_mut)]` on by default

    Finished test [unoptimized + debuginfo] target(s) in 1.53s
     Running target/debug/deps/solana_runtime-e554a1125996be96

running 4 tests
test accounts_index::tests::test_purge ... ok
test accounts_db::tests::test_purge_slot_after_root ... ok
test accounts_db::tests::test_purge_slot_not_root ... ok
[2019-12-11T02:35:10.944116602Z INFO  solana_metrics::metrics] metrics disabled: SOLANA_METRICS_CONFIG: environment variable not found
[2019-12-11T02:35:11.036643738Z INFO  solana_runtime::accounts_db] accounts_db: total_stores: 11 newest_slot: 10 oldest_slot: 0 max_slot: 0 (num=1) min_slot: 0 (num=1)
[2019-12-11T02:35:11.037734425Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=11i
[2019-12-11T02:35:11.051088246Z INFO  solana_runtime::accounts_db] accounts_db: total_stores: 6 newest_slot: 12 oldest_slot: 0 max_slot: 0 (num=1) min_slot: 0 (num=1)
[2019-12-11T02:35:11.053194689Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=6i
[2019-12-11T02:35:11.053752700Z INFO  solana_runtime::accounts_db] accounts_db: total_stores: 5 newest_slot: 12 oldest_slot: 0 max_slot: 0 (num=1) min_slot: 0 (num=1)
[2019-12-11T02:35:11.054017365Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=5i
[2019-12-11T02:35:11.059109835Z INFO  solana_runtime::accounts_db] accounts_db: total_stores: 5 newest_slot: 12 oldest_slot: 0 max_slot: 0 (num=1) min_slot: 0 (num=1)
[2019-12-11T02:35:11.059451173Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=5i
test bank::tests::test_purge_empty_accounts ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 200 filtered out

```

Reply:
Oops, https://github.com/solana-labs/solana/pull/7408/files#diff-2099c5256db4eb5975c8834af38f6456R1931 correctly triggers an assertion.